ledger/catchupaccessor.go

```diff
@@ -0,0 +1,490 @@
+	// verify it's parsable :
+	_, _, err = ParseCatchpointLabel(label)
+	if err != nil {
+		return
+	}
+	_, err = c.accountsq.writeCatchpointStateString(ctx, "catchpointCatchupLabel", label)
+	if err != nil {
+		return fmt.Errorf("unable to write catchpoint catchup state 'catchpointCatchupLabel': %v", err)
+	}
+	return
+}
+
+// ResetStagingBalances resets the current staging balances, preparing for a new set of balances to be added
+func (c *CatchpointCatchupAccessor) ResetStagingBalances(ctx context.Context, newCatchup bool) (err error) {
+	wdb := c.ledger.trackerDB().wdb
+	err = wdb.Atomic(func(tx *sql.Tx) (err error) {
+		err = resetCatchpointStagingBalances(ctx, tx, newCatchup)
+		if err != nil {
+			return fmt.Errorf("unable to reset catchpoint catchup balances : %v", err)
+		}
+		if !newCatchup {
+			sq, err := accountsDbInit(tx, tx)
+			if err != nil {
+				return fmt.Errorf("unable to initialize accountsDbInit: %v", err)
+			}
+			_, err = sq.writeCatchpointStateUint64(ctx, "catchpointCatchupBalancesRound", 0)
+			if err != nil {
+				return err
+			}
+
+			_, err = sq.writeCatchpointStateUint64(ctx, "catchpointCatchupBlockRound", 0)
+			if err != nil {
+				return err
+			}
+
+			_, err = sq.writeCatchpointStateString(ctx, "catchpointCatchupLabel", "")
+			if err != nil {
+				return err
+			}
+			_, err = sq.writeCatchpointStateUint64(ctx, "catchpointCatchupState", 0)
+			if err != nil {
+				return fmt.Errorf("unable to write catchpoint catchup state 'catchpointCatchupState': %v", err)
+			}
+		}
+		return
+	})
+	return
+}
+
+// CatchpointCatchupAccessorProgress is used by the caller of ProgressStagingBalances to obtain progress information
+type CatchpointCatchupAccessorProgress struct {
+	TotalAccounts     uint64
+	ProcessedAccounts uint64
+	ProcessedBytes    uint64
+	TotalChunks       uint64
+	SeenHeader        bool
+}
+
+// ProgressStagingBalances deserialize the given bytes as a temporary staging balances
+func (c *CatchpointCatchupAccessor) ProgressStagingBalances(ctx context.Context, sectionName string, bytes []byte, progress *CatchpointCatchupAccessorProgress) (err error) {
+	if sectionName == "content.msgpack" {
+		return c.processStagingContent(ctx, bytes, progress)
+	}
+	if strings.HasPrefix(sectionName, "balances.") && strings.HasSuffix(sectionName, ".msgpack") {
+		return c.processStagingBalances(ctx, bytes, progress)
+	}
+	// we want to allow undefined sections to support backward compatibility.
+	c.log.Warnf("CatchpointCatchupAccessor::ProgressStagingBalances encountered unexpected section name '%s' of length %d, which would be ignored", sectionName, len(bytes))
+	return nil
+}
+
+// processStagingContent deserialize the given bytes as a temporary staging balances content
+func (c *CatchpointCatchupAccessor) processStagingContent(ctx context.Context, bytes []byte, progress *CatchpointCatchupAccessorProgress) (err error) {
+	if progress.SeenHeader {
+		return fmt.Errorf("content chunk already seen")
```

Comment:
Shall we add context to the error message?

Reply:
sure.